src/url_aggregator.cpp

```diff
@@ -54,18 +114,43 @@ bool url_aggregator::set_hostname(const std::string_view input) {
 }
 
 [[nodiscard]] std::string url_aggregator::get_origin() const noexcept {
-  // TODO: Implement this
+  if (is_special()) {
+    // Return a new opaque origin.
+    if (type == scheme::FILE) { return "null"; }
+
+    return helpers::concat(get_protocol(), "//", get_host());
+  }
+
+  if (checkers::begins_with(get_protocol(), "blob")) {
+    std::string_view path = retrieve_base_pathname();
+    if (path.length() > 0) {
+      ada::result<ada::url> path_result = ada::parse<ada::url>(path);
+      if (path_result) {
+        if (path_result->is_special()) {
+          return path_result->get_protocol() + "//" + path_result->get_host();
+        }
+      }
+    }
+  }
+
+  // Return a new opaque origin.
   return "null";
 }
 
 [[nodiscard]] std::string_view url_aggregator::get_username() const noexcept {
-  // TODO: Implement this properly
-  return helpers::substring(buffer, components.protocol_end, components.username_end);
+  bool has_authority = checkers::begins_with(buffer.substr(components.protocol_end, 3), "://");
```

Comment:
Hmmm. This creates a temporary substring needlessly.

something like this ought to do it... (to be tested and verified)

```suggestion
  bool has_authority = (components.protocol_end + 3 <= buffer.size()) && helpers::substring(buffer, components.protocol_end,  components.protocol_end + 3) == "://";
```